runtime/lua/vim/ui.lua

```diff
@@ -1,5 +1,217 @@
+---
+---Providing an implementation
+---
+---nvim comes with three providers that interface with various
+---graphics protocols supported by |TUI|s and GUIs.
+---
+---1. kitty
+---2. iterm2
+---3. sixel
+---
+---The global provider to use with all placements is specified via the option,
+---*imgprovider*, which is set to *kitty* by default. When the global provider
+---is changed, the old provider is unloaded.
+---
+---```lua
+---vim.o.imgprovider = 'kitty'
+---```
+---
+---The image API also supports 3rd party provider integrations. To create a
+---new provider, the implementer should leverage `vim.ui.img.providers.new()`.
+---
+---To register the provider such that it is accessible globally, the implementer
+---should assign it with a name to the dictionary at `vim.ui.img.providers`.
+---
+---```lua
+---vim.ui.img.providers['neovide'] = vim.ui.img.providers.new({
+---    ---(Optional) Called to initialize the provider.
+---    ---@param ... any arguments for the specific provider upon loading
+---    load = function(...)
+---      -- Implement here
+---    end,
+---
+---    ---(Optional) Called to cleanup the provider.
+---    unload = function()
+---      -- Implement here
+---    end,
+---
+---    ---(Optional) Reports whether this provider is supported in the current environment.
+---    ---@param on_supported? fun(supported:boolean) callback when finished checking
+---    supported = function(on_supported)
+---        -- Implement here
+---    end,
+---
+---    ---Displays an image, returning (through callback) an id tied to the instance.
+---    ---@param img vim.ui.Image image data container to display
+---    ---@param opts vim.ui.img.Opts specification of how to display the image
+---    ---@param on_shown? fun(err:string|nil, id:integer|nil) callback when finished showing
+---    ---@return integer id unique identifier connected to the displayed image (not vim.ui.Image)
+---    show = function(img, opts, on_shown)
+---        -- Implement here
+---    end,
+---
+---    ---Hides one or more displayed images.
+---    ---@param ids integer[] list of displayed image ids to hide
+---    ---@param on_hidden fun(err:string|nil, ids:integer[]|nil) callback when finished hiding
+---    hide = function(ids, on_hidden)
+---        -- Implement here
+---    end,
+---
+---    ---(Optional) Updates an image, returning (through callback) a refreshed id tied to the instance.
+---    ---If not specified, nvim will invoke `hide(id)` followed by `show(img, opts, on_updated)`.
+---    ---@param id integer id of the displayed image to update
+---    ---@param opts vim.ui.img.Opts specification of how to display the image
+---    ---@param on_updated? fun(err:string|nil, id:integer|nil) callback when finished updating
+---    ---@return integer id unique identifier connected to the displayed image (not vim.ui.Image)
+---    update = function(id, opts, on_updated)
+---        -- Implement here
+---    end,
+---})
+---```
+M.img = require('vim.ui.img')
```

Comment:
`_defer_require` might be useful here https://github.com/neovim/neovim/blob/dfad6138131f86a229d6fb9fb21742a8b979ede7/runtime/lua/vim/lsp.lua#L4